File: 프로그래머스/1/172928. 공원 산책/공원 산책.js
function solution(park, routes) {
    const R = park.length;
    const C = park[0].length;
    let r = 0;
    let c = 0;
    const dr = [-1, 0, 1, 0];
    const dc = [0, 1, 0, -1];
    const dirIdxMap = {
        'N': 0,
        'E': 1,
        'S': 2,
        'W': 3,
    }
    
    while (park[r][c] !== 'S') {
        if (c < C - 1) {
            c++;
        } else {
            r++;
            c = 0;
        }
    }
    
    for (let route of routes) {
        const inst = route.split(' ');
        const dirIdx = dirIdxMap[inst[0]];
        const dist = +inst[1];
        let cr = r;
        let cc = c;
        let isValid = true;
        
        for (let i = 1; i <= dist; i++) {
            cr += dr[dirIdx];
            cc += dc[dirIdx];
            
            if (cr < 0 || cr >= R
                || cc < 0 || cc >= C
                || park[cr][cc] === 'X'
               ) {
                isValid = false;
                break;
            }
        }
        
        if (!isValid) continue;
        
        r = cr;
        c = cc;
    }
    
    return [r, c];
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    H = len(park)
    W = len(park[0])
    park_map = []
    D = {
        "E": [0, 1],
        "W": [0, -1],
        "S": [1, 0],
        "N": [-1, 0]
    }
    
    for h in range(H):
        row = []
        for w in range(W):
            x = park[h][w]
            
            if x == 'S':
                ch, cw = h, w
            
            row.append(x)
        park_map.append(row)
    
    for route in routes:
        d, n = route.split()
        nh, nw = ch, cw
        
        for _ in range(int(n)):
            nh += D[d][0]
            nw += D[d][1]
            
            if (nh < 0 or nh >= H) or (nw < 0 or nw >= W) or park_map[nh][nw] == 'X':
                nh, nw = ch, cw
                break
            
        ch, cw = nh, nw
        
    return ch, cw
        